rustls-libssl/src/lib.rs

```diff
@@ -527,23 +536,99 @@ impl Ssl {
         let client_conn = ClientConnection::new(Arc::new(config), sni_server_name.clone())
             .map_err(error::Error::from_rustls)?;
 
-        self.conn = Some(client_conn.into());
+        self.conn = ConnState::Client(client_conn.into(), verifier);
+        Ok(())
+    }
+
+    fn accept(&mut self) -> Result<(), error::Error> {
+        self.set_server_mode();
```

Comment:
Unfortunately because `void SSL_set_accept_state(SSL *ssl)` (& similarly `SSL_set_connect_state`) cannot report errors, I think this we can't make those fallible. The openssl versions literally splat any and all existing connection state, even if called mid- or post-handshake!

However, I will branch on the existing `self.mode` value here (and in `SSL_connect`); that makes `SSL_accept` and `SSL_connect` equivalent to `SSL_do_handshake` if the connection mode is already set, and matches the openssl behaviour. (nb. `SSL_do_handshake` is in a later commit than this PR.)

Reply:
Sounds reasonable to me